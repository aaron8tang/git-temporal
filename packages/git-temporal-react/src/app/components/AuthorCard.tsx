import React from 'react';
import { style } from 'app/styles';

import { AuthorGravatarImage } from 'app/components/AuthorGravatarImage';
import { AddedDeleted } from 'app/components/AddedDeleted';
import { PercentBar } from 'app/components/PercentBar';
import { EpochSpan } from 'app/components/EpochSpan';
import { Selectable } from 'app/components/Selectable';

import { IAuthorStats } from 'app/interfaces';

export interface AuthorCardProps {
  author: IAuthorStats;
  totalLinesAdded: number;
  totalLinesDeleted: number;
  totalCommits: number;
  maxImpact: number;
  maxCommits: number;
  onClick: (evt, commit: IAuthorStats, index: number) => void;
  isHighlighted?: boolean;
  index?: number;
  style?: string | object;
}

const menuItemStyle = {
  fontSize: 'initial',
  display: 'flex',
  padding: 10,
  marginRight: -10,
  width: '100%',
};

const identifiersStyle = {
  _extends: 'flexColumn',
  width: '100%',
  marginLeft: 10,
};
const authorNameStyle = {
  marginBottom: 5,
};

const addedDeletedStyle = {
  marginTop: 6,
};

export const AuthorCard = (props: AuthorCardProps): JSX.Element => {
  const { author, maxImpact, maxCommits, index, isHighlighted } = props;
  const addStyles = isHighlighted ? 'selected' : {};

  return (
    <div style={style('flexRow', props.style)}>
      <Selectable
        style={style(menuItemStyle, addStyles)}
        onClick={evt => props.onClick(evt, author, index)}
      >
        <AuthorGravatarImage emails={author.authorEmails} />
        <div style={style(identifiersStyle)}>
          <div
            title={author.authorEmails.join(', ')}
            style={style(authorNameStyle)}
          >
            {author.authorName}
          </div>
          <div>
            <span style={style('smallerText')}>
              <EpochSpan
                firstEpochTime={author.lastCommitOn}
                secondEpochTime={Date.now() / 1000}
              />
              <span> ago</span>
            </span>
            <AddedDeleted
              linesAdded={author.linesAdded}
              linesDeleted={author.linesDeleted}
              style={style(addedDeletedStyle)}
            />
          </div>
          <PercentBar
            numerator={author.linesAdded + author.linesDeleted}
            denominator={maxImpact}
          />
          <div style={style('smallerText')}>
            {author.commits.length}
            <span> commits spanning </span>
            <EpochSpan
              firstEpochTime={author.firstCommitOn}
              secondEpochTime={author.lastCommitOn}
            />
          </div>
          <PercentBar
            numerator={author.commits.length}
            denominator={maxCommits}
          />
        </div>
      </Selectable>
    </div>
  );
};
