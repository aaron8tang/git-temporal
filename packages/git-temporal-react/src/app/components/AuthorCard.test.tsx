import React from 'react';
import { shallow } from 'enzyme';

import { AuthorCard } from './AuthorCard';

const testProps = {
  author: {
    authorName: 'Zaphod Beeblebrox',
    authorEmails: ['president@galaxy.gov'],
    linesAdded: 100,
    linesDeleted: 30,
    totalCommits: 70,
    firstCommitOn: 132344525,
    lastCommitOn: 132344555,
  },
  totalLinesAdded: 200,
  totalLinesDeleted: 60,
  totalCommits: 140,
  maxImpact: 260,
  maxCommits: 140,
};

describe('components/AuthorCard', () => {
  describe('when rendered with required props', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallow(<AuthorCard {...testProps} />);
    });

    test('it should match snapshot (bars should be at 50%)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
