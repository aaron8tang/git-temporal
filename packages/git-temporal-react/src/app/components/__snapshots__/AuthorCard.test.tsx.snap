// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AuthorCard when rendered with required props it should match snapshot (bars should be at 50%) 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <Selectable
    onClick={[Function]}
    style={
      Object {
        "display": "flex",
        "fontSize": "initial",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <AuthorGravatarImage
      emails={
        Array [
          "president@galaxy.gov",
        ]
      }
    />
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "marginLeft": 10,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "marginBottom": 5,
          }
        }
        title="president@galaxy.gov"
      >
        Zaphod Beeblebrox
      </div>
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 11,
            "fontWeight": "normal",
            "lineHeight": 1.35,
          }
        }
      >
        <div>
          5
          <span>
             commits spanning 
          </span>
          <EpochSpan
            firstEpochTime={132344525}
            secondEpochTime={132344555}
          />
        </div>
      </div>
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 11,
            "fontWeight": "normal",
            "lineHeight": 1.35,
          }
        }
      >
        <span>
          Last commit 
        </span>
        <EpochSpan
          firstEpochTime={132344555}
          secondEpochTime={1539563458}
        />
        <span>
           ago
        </span>
      </div>
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 11,
            "fontWeight": "normal",
            "lineHeight": 1.35,
          }
        }
      >
        <AddedDeleted
          linesAdded={100}
          linesDeleted={30}
          style={
            Object {
              "float": "none",
            }
          }
        />
        <span>
           lines changed 
        </span>
      </div>
    </div>
  </Selectable>
</div>
`;
